Wrappers/Python/cil/optimisation/functions/TotalGeneralisedVariation.py

```diff
@@ -0,0 +1,211 @@
+        self.warmstart  = warmstart
+        if self.warmstart:
+            self.hasstarted = False        
+
+
+    def __call__(self, x):
+        
+        if not hasattr(self, 'pdhg'):   
+            return 0.0        
+        else:              
+            # Compute alpha1 * || Du - w || + alpha0 * ||Ew||, 
+            # where (u,w) are solutions coming from the proximal method below.
+            tmp = self.f(self.pdhg.operator.direct(self.pdhg.solution))
+            return tmp
+
+    def _setup_solver(self, x, tau=1.0):
+
+        # tau_sqrt = np.sqrt(tau)
+        
+        if not hasattr(self, 'domain'):
+            
+            # sirf compatibility
+            try:
+                self.domain = x.geometry
+            except:
+                self.domain = x            
+                        
+        if not hasattr(self, 'operator'):
+            
+            self.Gradient = GradientOperator(self.domain, correlation = self.correlation, backend = self.backend)  
+            self.SymGradient = SymmetrisedGradientOperator(self.Gradient.range, correlation = self.correlation, backend = self.backend)  
+            self.ZeroOperator = ZeroOperator(self.domain, self.SymGradient.range)
+            self.IdentityOperator = - IdentityOperator(self.Gradient.range)
+
+            #    BlockOperator = [ Gradient      - Identity  ]
+            #                    [ ZeroOperator   SymGradient] 
+            self.operator = BlockOperator(self.Gradient, self.IdentityOperator, 
+                                          self.ZeroOperator,self.SymGradient,shape=(2,2)) 
+
+        if not all(hasattr(self, attr) for attr in ["g"]):
+            self.g1 = 0.5*L2NormSquared(b = x)
+            self.g2 = ZeroFunction()               
+            self.g = BlockFunction(self.g1, self.g2)
+
+
+        if not all(hasattr(self, attr) for attr in ["f"]):
+            # parameters to set up PDHG algorithm
+            self.f1 = tau*self.alpha1 * MixedL21Norm()
+            self.f2 = tau*self.alpha0 * MixedL21Norm()
+            self.f = BlockFunction(self.f1, self.f2)  
+         
+        if self.warmstart:     
+            if self.hasstarted:
+                tmp_initial = self.pdhg.solution
+            else:
+                tmp_initial = None
+                self.hasstarted = True 
+        else:
+            tmp_initial = None
+
+        # setup PDHG                           
+        self.pdhg = PDHG(initial = tmp_initial, f = self.f, g=self.g, operator = self.operator,
+                update_objective_interval = self.update_objective_interval,
+                max_iteration = self.iterations) 
+        self.pdhg.run(verbose=self.verbose) 
+
+                                     
+    def proximal(self, x, tau = 1.0, out = None):
+        
+        self._setup_solver(x, tau)  
+                    
+        if out is None:
+            return self.pdhg.solution[0]
+        else:
+            out.fill(self.pdhg.solution[0])            
```

Comment:
the member `pdhg` seems to be only used to access its member `solution[0]`. Since any PDHG instance has a large memory footprint, one could actually save just the solution in case that the user sets `warm_start=True`

This method could look like

```python

def proximal(self, x, tau=1.0, out=None):
    # after this the pdhg algorithm is garbage collected
    tmp_out = self._setup_and_run_solver(x, tau)
    if self.warm_start:
        # save solver solution to be used in _setup_and_run_solver
        # TODO modify _setup_and_run_solver accordingly
        self.tmp_initial = tmp_out
    if out is None:
        return tmp_out[0]
    else:
        out.fill(tmp_out[0])
```